cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -0,0 +1,213 @@
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        settableFuture.setException(t);
+      }
+    }, Threads.SAME_THREAD_EXECUTOR);
+
+    try {
+      settableFuture.get(ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (ExecutionException | InterruptedException | TimeoutException ex) {
+      if (ex.getCause() instanceof KeeperException.NoNodeException) {
+        throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+      }
+      throw Throwables.propagate(ex);
+    }
+  }
+
+  @Override
+  public RouteConfig fetch(final ProgramId serviceId) throws NotFoundException {
+    if (routeConfigMap.putIfAbsent(serviceId, SettableFuture.<RouteConfig>create()) != null) {
+      return getConfig(serviceId, routeConfigMap.get(serviceId));
+    }
+    Future<RouteConfig> future = getAndWatchData(serviceId, new ZKRouteWatcher(serviceId));
+    return getConfig(serviceId, future);
+  }
+
+  private RouteConfig getConfig(ProgramId serviceId, Future<RouteConfig> future) throws NotFoundException {
+    try {
+      return future.get(ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      if (e.getCause() instanceof KeeperException.NoNodeException) {
+        throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+      }
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private static String getZKPath(ProgramId serviceId) {
+    return String.format("/routestore/%s", ServiceDiscoverable.getName(serviceId));
+  }
+
+  static final Codec<RouteConfig> ROUTE_CONFIG_CODEC = new Codec<RouteConfig>() {
+
+    @Override
+    public byte[] encode(RouteConfig object) throws IOException {
+      return Bytes.toBytes(GSON.toJson(object.getRoutes()));
+    }
+
+    @Override
+    public RouteConfig decode(byte[] data) throws IOException {
+      Map<String, Integer> routes = GSON.fromJson(Bytes.toString(data), MAP_STRING_INTEGER_TYPE);
+      return new RouteConfig(routes);
+    }
+  };
+
+  private Future<RouteConfig> getAndWatchData(final ProgramId serviceId, Watcher watcher) {
+    OperationFuture<NodeData> future = zkClient.getData(getZKPath(serviceId), watcher);
+    final SettableFuture<RouteConfig> settableFuture = SettableFuture.create();
+    Futures.addCallback(future, new FutureCallback<NodeData>() {
+      @Override
+      public void onSuccess(NodeData result) {
+        try {
+          RouteConfig route = ROUTE_CONFIG_CODEC.decode(result.getData());
+          settableFuture.set(route);
+          routeConfigMap.put(serviceId, settableFuture);
+        } catch (IOException ex) {
+          LOG.debug("Unable to deserialize the config for service {}. Got data {}", serviceId, result.getData());
+          settableFuture.setException(ex);
+        }
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        settableFuture.setException(t);
```

Comment:
We should also remove the future from the map, right? Otherwise future calls will continue to use this future and think there is an exception. Same for `IOException` in `onSuccess`